### PR TITLE
Chap2

### PR DESCRIPTION
**Pull request by specdrake**
### Solutions for Chapter 2

cc @vrom911 @chshersh

### CODE REVIEW

File: learn4haskell.cabal [hunk 1]
```diff
@@ -22,7 +22,7 @@ source-repository head
   location:            https://github.com/kowainik/learn4haskell.git
 
 common common-options
-  build-depends:       base ^>= 4.14.0.0
+  build-depends:       base ^>= 4.13.0.0
```
Comment (vrom911) [inline learn4haskell.cabal]: In order to tests to pass, you can not restrict it having the upper version, that is needed for our CI system
```suggestion
  build-depends:       base >= 4.13.0.0 && < 4.15
```

Do you have any particular reason for this change? Does any of our instruction didn't work for you?

File: src/Chapter2.hs [hunk 2]
```diff
@@ -620,7 +640,10 @@ Write a function that takes elements of a list only on even positions.
 >>> takeEven [2, 1, 3, 5, 4]
 [2,3,4]
 -}
-takeEven = error "takeEven: Not implemented!"
+takeEven :: [a] -> [a]
+takeEven [] = []
+takeEven (x:_:xs) = x : takeEven xs
+takeEven (x:_) = [x]
```
Comment (vrom911) [inline src/Chapter2.hs:646]: Great pattern matching skills 👏🏼

File: src/Chapter2.hs [hunk 3]
```diff
@@ -727,7 +750,8 @@ value of the element itself
 🕯 HINT: Use combination of 'map' and 'replicate'
 -}
 smartReplicate :: [Int] -> [Int]
-smartReplicate l = error "smartReplicate: Not implemented!"
+-- smartReplicate = concat . map (\x -> replicate x x)
+smartReplicate = ((\x -> replicate x x) =<<)
```
Comment (vrom911) [inline src/Chapter2.hs:754]: Interesting usage of the `Monad instance in here 🔆 

`concatMap` is also one functions, so you could use it in here. It is usually a bit  more readable 🙂

File: src/Chapter2.hs [hunk 4]
```diff
@@ -740,7 +764,9 @@ the list with only those lists that contain a passed element.
 
 🕯 HINT: Use the 'elem' function to check whether an element belongs to a list
 -}
-contains = error "contains: Not implemented!"
+contains :: Int -> [[Int]] -> [[Int]]
+-- contains n = foldl (\xss xs -> if (elem n xs) then xs:xss else xss) []
+contains n = filter (elem n)
```
Comment (vrom911) [inline src/Chapter2.hs:769]: Yep, this solution is just 👌🏼

File: src/Chapter2.hs [hunk 5]
```diff
@@ -780,13 +806,15 @@ Let's now try to eta-reduce some of the functions and ensure that we
...
 
 -- TODO: type ;)
-listElementsLessThan x l = filter (< x) l
+listElementsLessThan :: Ord a => a -> [a] -> [a]
+listElementsLessThan x = filter (< x)
 
 -- Can you eta-reduce this one???
-pairMul xs ys = zipWith (*) xs ys
+pairMul :: Num a => [a] -> [a] -> [a]
+pairMul = zipWith (*)
```
Comment (vrom911) [inline src/Chapter2.hs:817]: Eta-reduction award 🏆

File: src/Chapter2.hs [hunk 6]
```diff
@@ -857,7 +889,9 @@ and reverses it.
   function, but in this task, you need to implement it manually. No
   cheating!
 -}
-rewind = error "rewind: Not Implemented!"
+-- rewind [] = []
+rewind :: [a] -> [a]
+rewind = foldl (flip (:)) []
```
Comment (vrom911) [inline src/Chapter2.hs:894]: Wow, that's a clever solution using standard functions! 🍸 

And you correctly noticed that you don't need to handle the case with `[]` 🙂

**Review by vrom911 — APPROVED** [review]
Your solutions look awesome! Well done with Chapter Two 🕺🏼 


Btw, let me know if I could help with the `base` version issue 🙂

File: learn4haskell.cabal (hunk 1 above)
Comment (specdrake) [reply learn4haskell.cabal]: Yeah, cabal doesn't build with 4.14 version, so I had to use 4.13 in order to conduct chapter tests. I am using ghc version 8.8.4 globally. (I had installed it through ghcup)

**Comment by specdrake** [comment]
I really appreciate your learn4haskell initiative. It's really helpful for both Haskell and open-source beginners. Keep up the great work!

File: learn4haskell.cabal (hunk 1 above)
Comment (vrom911) [reply learn4haskell.cabal]: Oh I see. Yes, indeed, 8.8.4 requires `4.13` 🙂 We are recommending to use the latest `8.10` in course, that is why we put these restrictions 👍🏼 Glad that is working for you!

**Pull request merged**